flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/RoundRobinOperatorStateRepartitioner.java

```diff
@@ -31,36 +32,59 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 
 /**
  * Current default implementation of {@link OperatorStateRepartitioner} that redistributes state in round robin fashion.
  */
+@Internal
 public class RoundRobinOperatorStateRepartitioner implements OperatorStateRepartitioner {
 
 	public static final OperatorStateRepartitioner INSTANCE = new RoundRobinOperatorStateRepartitioner();
 	private static final boolean OPTIMIZE_MEMORY_USE = false;
 
 	@Override
 	public List<List<OperatorStateHandle>> repartitionState(
-			List<OperatorStateHandle> previousParallelSubtaskStates,
+			List<List<OperatorStateHandle>> previousParallelSubtaskStates,
+			int oldParallelism,
 			int newParallelism) {
 
 		Preconditions.checkNotNull(previousParallelSubtaskStates);
 		Preconditions.checkArgument(newParallelism > 0);
-
-		// Reorganize: group by (State Name -> StreamStateHandle + Offsets)
-		GroupByStateNameResults nameToStateByMode = groupByStateName(previousParallelSubtaskStates);
-
-		if (OPTIMIZE_MEMORY_USE) {
-			previousParallelSubtaskStates.clear(); // free for GC at to cost that old handles are no longer available
-		}
+		Preconditions.checkArgument(previousParallelSubtaskStates.size() == oldParallelism,
+			"This method still depends on the order of the new and old operators");
 
 		// Assemble result from all merge maps
 		List<List<OperatorStateHandle>> result = new ArrayList<>(newParallelism);
 
-		// Do the actual repartitioning for all named states
-		List<Map<StreamStateHandle, OperatorStateHandle>> mergeMapList =
+		List<Map<StreamStateHandle, OperatorStateHandle>> mergeMapList;
+
+		// We only round-robin repartition UNION state if new parallelism equals to the old one.
+		if (newParallelism == oldParallelism) {
+			Map<String, List<Tuple2<StreamStateHandle, OperatorStateHandle.StateMetaInfo>>> unionStates = collectUnionStates(previousParallelSubtaskStates);
+
+			if (unionStates.isEmpty()) {
```

Comment:
I give the assumption that each operator actually get the same BROADCAST state when they are running. Please correct me if I am wrong, previous method  should just get `index%old parallelism` BROADCAST state.

Reply:
Yes, the idea behind broadcast state is that only one parallel instance needs to checkpoint it, because it is assumed to be the same on all instances. Then on restore, all instances can restore from the same state.

Reply:
In some sense, it is like a union state that enforces that only one subtask reports state that is supposed to be the same on all subtasks.

Reply:
From "All tasks checkpoint their broadcast state" section in [broadcast state document](https://ci.apache.org/projects/flink/flink-docs-stable/dev/stream/state/broadcast_state.html#important-considerations) said, I think we should not consider broadcast state repartition problem.

Reply:
Seems that you are right. Sorry for that, my mistake. Then please ignore my comments about that, it seems this was different from what I thought.